orders-accounting/src/cache.rs

```diff
@@ -0,0 +1,258 @@
+        }
+        self.data.give_balances.sub_unsigned(undo.id, undo.give_balance)?;
+
+        Ok(())
+    }
+
+    fn undo_withdraw_order(&mut self, undo: WithdrawOrderUndo) -> Result<()> {
+        ensure!(
+            self.get_order_data(&undo.id)?.is_none(),
+            Error::InvariantOrderDataExistForWithdrawUndo(undo.id)
+        );
+        self.data.order_data.undo_merge_delta_data_element(undo.id, undo.undo_data)?;
+
+        ensure!(
+            self.get_ask_balance(&undo.id)?.unwrap_or(Amount::ZERO) == Amount::ZERO,
+            Error::InvariantOrderAskBalanceExistForWithdrawUndo(undo.id)
+        );
+        self.data.ask_balances.add_unsigned(undo.id, undo.ask_balance)?;
+
+        ensure!(
+            self.get_give_balance(&undo.id)?.unwrap_or(Amount::ZERO) == Amount::ZERO,
+            Error::InvariantOrderGiveBalanceExistForWithdrawUndo(undo.id)
+        );
+        self.data.give_balances.add_unsigned(undo.id, undo.give_balance)?;
+
+        Ok(())
+    }
+
+    fn undo_fill_order(&mut self, undo: FillOrderUndo) -> Result<()> {
+        if let Some(undo_data) = undo.undo_data {
+            ensure!(
+                self.get_order_data(&undo.id)?.is_none(),
+                Error::InvariantOrderDataExistForWithdrawUndo(undo.id)
+            );
+            self.data.order_data.undo_merge_delta_data_element(undo.id, undo_data)?;
+        }
+
+        self.data.ask_balances.add_unsigned(undo.id, undo.ask_balance)?;
+        self.data.give_balances.add_unsigned(undo.id, undo.give_balance)?;
+
+        Ok(())
+    }
+}
+
+impl<P: OrdersAccountingView> OrdersAccountingView for OrdersAccountingCache<P> {
+    type Error = Error;
+
+    fn get_order_data(&self, id: &OrderId) -> Result<Option<OrderData>> {
+        match self.data.order_data.get_data(id) {
+            accounting::GetDataResult::Present(d) => Ok(Some(d.clone())),
+            accounting::GetDataResult::Deleted => Ok(None),
+            accounting::GetDataResult::Missing => {
+                Ok(self.parent.get_order_data(id).map_err(|_| Error::ViewFail)?)
+            }
+        }
+    }
+
+    fn get_ask_balance(&self, id: &OrderId) -> Result<Option<Amount>> {
+        let parent_supply = self.parent.get_ask_balance(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.ask_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+
+    fn get_give_balance(&self, id: &OrderId) -> Result<Option<Amount>> {
+        let parent_supply = self.parent.get_give_balance(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.give_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+}
+
+impl<P: OrdersAccountingView> OrdersAccountingOperations for OrdersAccountingCache<P> {
+    fn create_order(&mut self, id: OrderId, data: OrderData) -> Result<OrdersAccountingUndo> {
```

Comment:
Reminder: Order id calculation should be automated + it should be enforced to use the automated results. We don't want people to be able to use any id they want.

Reply:
it is calculated in the tx-verifier based on tx input.

Reply:
Are you sure it's not possible to use a custom id? This is manually enforced for staking pools.

Reply:
It is possible to pass custom id to this function, e.g. in tests. But in transaction verifier it is called with id generated from utxo input.
For staking pool the situation is different because `TxInput::CreateStakePool(PoolId, _)` has pool id it it. So manual enforcement is required. But not for tokens and orders

Reply:
Please ensure this is tested, because abusing it can cause issues.

Reply:
Checked, it is tested in chainstate tests that an order is created with id produced from input0.